CI: Add aarch64 docker build and ciflow tags

snadampal opened the pull request:
adding workflows for aarch64 linux docker build with ACL installed as system dependency

atalman commented on .github/workflows/docker-builds.yml:
```diff
@@ -56,6 +56,7 @@ jobs:
           - docker-image-name: pytorch-linux-focal-linter
           - docker-image-name: pytorch-linux-jammy-cuda11.8-cudnn8-py3.9-linter
           - docker-image-name: pytorch-linux-jammy-py3-clang12-executorch
+          - docker-image-name: pytorch-linux-jammy-aarch64-py3.10-gcc11
```
To fix this docker build issue please use this approach: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations
with different runners required for different docker image then this could be passed to runs-on as matrix parameter.

snadampal replied:
extended matrix and included aarch64 linux job

malfet commented on .ci/docker/common/install_conda.sh:
```diff
@@ -9,6 +9,18 @@ if [ -n "$ANACONDA_PYTHON_VERSION" ]; then
   MINOR_PYTHON_VERSION=$(echo "$ANACONDA_PYTHON_VERSION" | cut -d . -f 2)
 
+if [[ $(uname -m) == "aarch64" ]]; then
+  BASE_URL="https://github.com/conda-forge/miniforge/releases/latest/download"
```
Why prefer miniforge rather than miniconda, which is more stable?

snadampal replied:
I see several packages don't exist on miniconda channel for aarch64. And also, we've been using the miniforge in the CD scripts.https://github.com/pytorch/builder/blob/main/aarch64_linux/aarch64_ci_setup.sh

malfet commented on .ci/docker/common/install_conda.sh:
```diff
@@ -51,12 +64,22 @@ if [ -n "$ANACONDA_PYTHON_VERSION" ]; then
-  if [ "$ANACONDA_PYTHON_VERSION" = "3.11" ] || [ "$ANACONDA_PYTHON_VERSION" = "3.12" ]; then
-    conda_install numpy=1.26.0 ${CONDA_COMMON_DEPS}
+  if [[ $(uname -m) == "aarch64" ]]; then
+    CONDA_COMMON_DEPS="astunparse pyyaml setuptools patchelf==0.17.2 pygit2==1.13.2 openblas==0.3.25=*openmp* ninja==1.11.1 scons==4.5.2"
```
Why do we need pygit2 or patchelf to build?
```suggestion
    CONDA_COMMON_DEPS="astunparse pyyaml setuptools openblas==0.3.25=*openmp* ninja==1.11.1 scons==4.5.2"
```

snadampal replied:
probably not because we are not packaging the wheels for external use. I will clean it up

snadampal replied:
removed pygit2 and patchelf.

malfet commented on .ci/docker/common/install_conda.sh:
```diff
@@ -51,12 +64,22 @@ if [ -n "$ANACONDA_PYTHON_VERSION" ]; then
   as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y python="$ANACONDA_PYTHON_VERSION"
 
   # Install PyTorch conda deps, as per https://github.com/pytorch/pytorch README
-  CONDA_COMMON_DEPS="astunparse pyyaml mkl=2021.4.0 mkl-include=2021.4.0 setuptools"
-  if [ "$ANACONDA_PYTHON_VERSION" = "3.11" ] || [ "$ANACONDA_PYTHON_VERSION" = "3.12" ]; then
-    conda_install numpy=1.26.0 ${CONDA_COMMON_DEPS}
+  if [[ $(uname -m) == "aarch64" ]]; then
+    CONDA_COMMON_DEPS="astunparse pyyaml setuptools patchelf==0.17.2 pygit2==1.13.2 openblas==0.3.25=*openmp* ninja==1.11.1 scons==4.5.2"
```
To avoid code duplication I would've rewrote it to something along those lines:
```
CONDA_COMMON_DEPS = "astunparse pyyaml setuptools"
# Pick appropriate LAPACK library for the platform
if [ $(uname -m) == "x86_64"]; then
    CONDA_COMMON_DEPS="${CONDA_COMMON_DEPS} mkl=2021.4.0 mkl-include=2021.4.0"
else
  CONDA_COMMON_DEPS="${CONDA_COMMON_DEPS} openblas==0.3.25=*openmp*  scons==4.5.2"
fi
```

malfet commented on .ci/docker/requirements-ci.txt:
```diff
@@ -248,7 +248,7 @@ tb-nightly==2.13.0a20230426
 #test that import:
 
-#typing-extensions
+typing-extensions
```
Why are we making this change?

snadampal replied:
this is required by pytorch builds on aarch64. I'm surprised why it's not required on x86.

malfet replied:
Do you mind submitting typing-extensions change separately (or give me a few min to search the reason behind the change)

snadampal replied:
I can submit separately. 
pytorch/requirement.txt lists this one:
https://github.com/pytorch/pytorch/blob/main/requirements.txt#L11

snadampal replied:
Hi @malfet ,  This is the [PR](https://github.com/pytorch/pytorch/pull/121136) I raised for "typing_extensions", also added why it is required.

snadampal replied:
the typing extensions PR is merged.

malfet commented on .ci/docker/requirements-ci.txt:
```diff
@@ -263,7 +263,8 @@ unittest-xml-reporting<=3.2.0,>=2.0.0
 #test that import:
 
-lintrunner==0.10.7
+#TODO: wheel not found on aarch64, and source build requires rust
+#lintrunner==0.10.7
```
Do not disable it, just constrain to x86_64 (according to https://peps.python.org/pep-0496/#strings )
```suggestion
#TODO: wheel not found on aarch64, and source build requires rust
lintrunner==0.10.7 ; platform_machine == "x86_64"
```

snadampal replied:
thank you, i wasn't aware of this syntax.

malfet commented on .ci/docker/ubuntu-aarch64/Dockerfile:
```diff
@@ -0,0 +1,74 @@
+ENV DOCS=$DOCS
+COPY requirements-ci.txt requirements-docs.txt /opt/conda/
+COPY ./common/install_conda.sh install_conda.sh
+COPY ./common/common_utils.sh common_utils.sh
+RUN bash ./install_conda.sh && rm install_conda.sh common_utils.sh /opt/conda/requirements-ci.txt /opt/conda/requirements-docs.txt
+RUN if [ -n "${UNINSTALL_DILL}" ]; then pip uninstall -y dill; fi
+
+# Install gcc
+ARG GCC_VERSION
+COPY ./common/install_gcc.sh install_gcc.sh
+RUN bash ./install_gcc.sh && rm install_gcc.sh
+
+# Install lcov for C++ code coverage
+COPY ./common/install_lcov.sh install_lcov.sh
+RUN  bash ./install_lcov.sh && rm install_lcov.sh
+
+# Build Arm Compute Library (ACL)
+COPY ./common/build_acl.sh build_acl.sh
+RUN bash ./build_acl.sh && rm build_acl.sh
+
+# (optional) Install database packages like LMDB and LevelDB
+ARG DB
+COPY ./common/install_db.sh install_db.sh
+RUN if [ -n "${DB}" ]; then bash ./install_db.sh; fi
+RUN rm install_db.sh
+ENV INSTALLED_DB ${DB}
+
+# (optional) Install vision packages like OpenCV and ffmpeg
+ARG VISION
+COPY ./common/install_vision.sh ./common/cache_vision_models.sh ./common/common_utils.sh ./
+RUN if [ -n "${VISION}" ]; then bash ./install_vision.sh; fi
+RUN rm install_vision.sh cache_vision_models.sh common_utils.sh
+ENV INSTALLED_VISION ${VISION}
+
+# Include BUILD_ENVIRONMENT environment variable in image
+ARG BUILD_ENVIRONMENT
+ENV BUILD_ENVIRONMENT ${BUILD_ENVIRONMENT}
+
+USER jenkins
+CMD ["bash"]
```
Why does it need to be a separate Docker file rather just adding a conditional ACL step to `ubuntu/Dockerfile`?

snadampal replied:
I have merged the required changes into the common dockerfile.

snadampal commented on .github/workflows/docker-builds.yml:
```diff
@@ -31,31 +31,51 @@ permissions: read-all
       matrix:
         include:
           - docker-image-name: pytorch-linux-focal-cuda12.1-cudnn8-py3-gcc9
+            runner: linux.12xlarge
```
I know repeating the same runner across all these jobs is not elegant. till i get the proper matrix generation working, i'm using it to test the aarch64 workflow.

snadampal replied:
updated the matrix generation logic in the next version.

atalman reviewed (approved):
LGTM the one failure is not related. @snadampal before landing please address @malfet comments

snadampal commented:
Hi @malfet , I have addressed your comments on this PR. and also raised a separate PR for "typing_extensions" package, here it is: https://github.com/pytorch/pytorch/pull/121136
Once the typing extensions PR is merged, I will rebase this PR

snadampal commented:
Hi @malfet , I have rebased the PR to the main, and all CI tests passed

malfet reviewed (changes requested):
The way PR is written right now, aarch64 workflow can not be tested until PR is landed, which makes me quite hesitant. Do you mind landing docker and probot changes in this PR and move aarch64.yml to the followup?

Two question about aarch64.yml:
 - Do we have enough compute capacity to run it on every trunk commit?
 - What is the time to signal (i.e. average build time and test time on aarch64 platform)
 - Is sccache used/ can we use sccache to speed up incremental builds?
If later is not the case (or say build+test takes 4+ hours) then I wonder if we should do opt-in first and/or perhaps make it part of periodic workflow

malfet commented on .github/workflows/docker-builds.yml:
```diff
@@ -31,31 +31,36 @@ permissions: read-all
     strategy:
       fail-fast: false
       matrix:
+        runner: [linux.12xlarge]
```
Please keep `self-hosted` here

snadampal replied:
it's added in line `63'

malfet commented on .github/workflows/docker-builds.yml:
```diff
@@ -31,31 +31,36 @@ permissions: read-all
       fail-fast: false
       matrix:
+        runner: [linux.12xlarge]
+        docker-image-name: [
+          pytorch-linux-focal-cuda12.1-cudnn8-py3-gcc9,
+          pytorch-linux-focal-cuda12.1-cudnn8-py3-gcc9-inductor-benchmarks,
+          pytorch-linux-focal-cuda11.8-cudnn8-py3-gcc9,
+          pytorch-linux-focal-py3.8-clang10,
+          pytorch-linux-focal-py3.11-clang10,
+          pytorch-linux-focal-py3.12-clang10,
+          pytorch-linux-focal-rocm-n-1-py3,
```
Is it still possible to stick to the block sequence stile, i.e.
`docker-image-name:
      - a 
      - b
      - c
```

snadampal committed:
CI: Add aarch64 linux ci workflow

adding workflows for aarch64 linux docker build and
pytorch wheel building with MKLDNN and ACL backend.

snadampal commented:
> The way PR is written right now, aarch64 workflow can not be tested until PR is landed, which makes me quite hesitant. Do you mind landing docker and probot changes in this PR and move aarch64.yml to the followup?
> 
> Two question about aarch64.yml:
> 
> * Do we have enough compute capacity to run it on every trunk commit?
> * What is the time to signal (i.e. average build time and test time on aarch64 platform)
> * Is sccache used/ can we use sccache to speed up incremental builds?
>   If later is not the case (or say build+test takes 4+ hours) then I wonder if we should do opt-in first and/or perhaps make it part of periodic workflow

Hi @malfet, thanks for the feedback.
* I have split the changes into 2 PRs, this is only for docker-build workflow now.
* and regarding the compute time and sccache usage, I will test it as part of linux-aarch64 workflow. 
* Currently the ARM64 linux runners are based on t4g.2xl, whereas for x86 linux is using .12xl. we need to migrate the ARM64 runner to c7g.4xl or even better if 16xl. 
* But to start with I will add tag based trigger, once the runners are upgraded, we can increase the frequency/scope.

malfet commented on .github/pytorch-probot.yml:
```diff
@@ -15,6 +15,7 @@ ciflow_push_tags:
 - ciflow/trunk
 - ciflow/unstable
 - ciflow/xpu
+- ciflow/aarch64
```
This should fix linter, though it makes it really hard how to test what is going on
```suggestion
```

malfet committed:
Update .github/pytorch-probot.yml

malfet commented:
@pytorchbot merge -f "Docker builds are green"
